src/services.py

```diff
@@ -0,0 +1,1033 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define DataHub services."""
+
+# Some General Notes
+# 1. `charm._state` does not behave in the standard Python way, you cannot get a reference
+#    to an object inside it and update it. For lasting updates you need to an assignment with
+#    the updated object.
+# 2. `process.wait()` causes some scripts (i.e. datahub-upgrade) to hang, which is why we use
+#    `process.wait_output()` even though we do not read the output. It does not have the same
+#    problem for some reason.
+# 3. Flags in `charm._state` have three states: True, False, None. Logic for them is very specific
+#    depending on the point in the program lifecycle and at certain places, explicit checks with
+#   `is` is required over general truthiness checks.
+
+import abc
+import logging
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, Optional
+
+import exceptions
+import literals
+import utils
+
+if TYPE_CHECKING:
+    from charm import DatahubK8SOperatorCharm  # noqa
+
+logger = logging.getLogger(__name__)
+
+
+def _kafka_topic_names(prefix: str) -> Optional[Dict[str, str]]:
```

Comment:
[Suggestion] You could create a util class containing such functions.

Reply:
It is exclusively used in this module, so I will leave it for now.